File: bowlingBall.cs
using System;
using System.Dynamic;

public class Test
{
	public static void Main()
	{
		// your code goes here
		int totalScore =0;
        // input for game
		int[][] balls = {
			new int[] { 10 },
    		new int[] { 9,1 },
    		new int[] { 5, 5 },
    		new int[] { 7, 2 },
    		new int[] { 10 },
    		new int[] { 10 },
    		new int[] { 10 },
    		new int[] { 9, 0 },
    		new int[] { 8, 2 },
    		new int[] { 9, 1, 10 }
		};
		Test t = new Test();
    	for(int i =0; i< balls.Length; i++) {
        	totalScore = t.BowlingScore(i, balls, totalScore);
        	Console.WriteLine(totalScore);
    	}
	}
	
    //return the total score for the ball user hit
	public int BowlingScore(int i, int[][] balls, int totalScore)
	{
		Test t = new Test();
    	dynamic currentScore = t.StrikeSpareScores(balls[i]);
    	if(i < balls.Length-1 && currentScore.strike == true) {
        	dynamic nextBallScore = t.StrikeSpareScores(balls[i+1]);
			
        	if(nextBallScore.strike == true && i+1 < (balls.Length -1) ) {
            	totalScore = totalScore + currentScore.score + nextBallScore.score + balls[i+2][0];
        	}
        	else {
            	totalScore = totalScore + currentScore.score + nextBallScore.score;
        	}
    	}
    	else if(i < balls.Length-1 && currentScore.spare == true) {
        	totalScore = totalScore + currentScore.score + balls[i+1][0];
		}
    	else {
        	totalScore = totalScore + currentScore.score;
    	}
    	return totalScore;
	}
	
    // returns an object having score for that ball, user has done strike or spare
	public object StrikeSpareScores(int[] ball)
	{
    	int score =0;
        bool strike = false;
        bool spare= false;

    	for(int i=0;i< ball.Length;i++) {
        	score = score + ball[i];
        	if(ball[0] == 10) {
            	strike = true;
        	}
        	if(score >= 10) {
            	spare = true;
        	}
    	}
    	dynamic result = new ExpandoObject();
    	result.score = score;
    	result.strike = strike;
    	result.spare = spare;
    	return result;
	}
}

File: bowlingBall.js
// balls input 
let balls = [[10],[9,1],[5,5],[7,2],[10],[10],[10],[9,0],[8,2],[9,1,10]];

/**
 * Bowling Ball Score
 * @function
 * @param {*} balls 
 * return total Score
 */
var bowlingBallScore = function(balls) {
    let totalScore =0;
    for(let i =0; i< balls.length; i++) {
        totalScore = bowlingScore(i, balls, totalScore);
        console.log(totalScore);
    }
    return totalScore;
}

let bowlingScore = function(i, balls, totalScore) {
    let currentScore = strikeSpareScores(balls[i]);
    if(i < balls.length-1 && currentScore.strike === true) {
        let nextBallScore = strikeSpareScores(balls[i+1]);
        if(nextBallScore.strike === true && i+1 < balls.length -1 ) {
            totalScore = totalScore + currentScore.score + nextBallScore.score + balls[i+2][0];
        }
        else {
            totalScore = totalScore + currentScore.score + nextBallScore.score;
        }
        
    }
    else if(i < balls.length-1 && currentScore.spare === true) {
        totalScore = totalScore + currentScore.score + balls[i+1][0];
    }
    else {
        totalScore = totalScore + currentScore.score;
    }
    return totalScore;
}

let strikeSpareScores = function(ball) {
    let score =0,
        strike = false,
        spare= false;

    for(let i=0;i< ball.length;i++) {
        score = score + ball[i];
        if(ball[0] === 10) {
            strike = true;
        }
        if(score >= 10) {
            spare = true;
        }
    }
    return {score,strike, spare};
}

let score = bowlingBallScore(balls);
